Make Data.Vect linear

mb64 opened the pull request:
This PR changes `Data.Vect.::` to be a linear constructor, like its `List` counterpart.

It further modifies many of the linear functions in `Data.Vect` to be annotated as such.

However, it falls short of making *everthing* that could be linear linear: higher-order functions remain nonlinear, to retain their use with nonlinear arguments.

I believe that although it adds some complexity, this change makes it overall conceptually simpler and much less surprising: one would expect that a function `{0 a : Type} -> (1 xs : Vect 4 a) -> Vect 3 a` would be impossible to write, but without this change, it's as easy as `tail`.  The goal of this change is to make `Vect` behave more "normally" wrt linearity.

ziman commented:
Nice!

gallais commented on libs/base/Data/Vect.idr:
```diff
@@ -181,9 +180,9 @@ merge = mergeBy compare
 public export
-reverse : Vect len elem -> Vect len elem
+reverse : (1 xs : Vect len elem) -> Vect len elem
 reverse xs = go [] xs
```
Worth noting that we can have a [proof-free implementation of `reverse`
using `foldl`](http://agda.github.io/agda-stdlib/Data.Vec.Base.html#7735).

mb64 replied:
While `foldr (flip cons)` would be great, it's again the issue that `foldr` and its higher-order friends only talk about unrestricted functions, so we have to implement the fold manually.

gallais commented on libs/base/Data/Vect.idr:
````diff
@@ -224,7 +223,7 @@ fromList' {len} ys (x::xs) =
 ||| ```
 public export
-fromList : (l : List elem) -> Vect (length l) elem
+fromList : (1 l : List elem) -> Vect (length l) elem
````
Why is this function so complicated?

```idris
fromList [] = []
fromList (x :: xs) = x :: fromList xs
```

ought to work! Consequently, I'm not sure what the use case for `fromList'` is.

mb64 replied:
It makes it tail recursive, which I would guess is more important in a strict language like Idris so that you won't bash your stack trying to convert an especially long list to a `Vect`.  Not certain tho – perhaps @edwinb can comment here.

ziman replied:
We've had this conversation about tail-recursive `public export` functions a few times recently ([#132](https://github.com/idris-lang/Idris2/issues/132#issuecomment-633285153), [#269](https://github.com/idris-lang/Idris2/pull/269#issuecomment-643168588)), so I'll just link what I'd repeat here. :)

There are issues with what I suggest, though, as @gallais pointed out, so take it just as an illustration why making everything tail recursive is not automatically a good thing.

gallais commented on libs/base/Data/Vect.idr:
```diff
@@ -780,9 +780,15 @@ range {len=S _} = FZ :: map FS range
+          -> (1 _ : Vect a (Vect b elem))
+          -> Vect a (Vect (S b) elem)
+    helper [] [] = []
+    helper (x::xs) (tl::tls) = (x::tl) :: helper xs tls
```
We cannot use `zipWith` but we could use applicative combinators.
Using Haskell notations, this is `(::) <$> xs <*> tls`.

gallais replied:
Alternatively we could come up with a naming scheme & provide
linear variants of the higher order functions. e.g. `zipWith1`.

mb64 replied:
Applicative combinators don't work either: `map` only maps unrestricted arrows to unrestricted arrows, and similary for `<*>`.  (I'm also not yet convinced that it *should* work in the general case – that is a question for someone with more knowledge of category theory than me.)

As for having separate linear variants for the higher-order functions, I like the idea, at least until Idris gets multiplicity polymorphism (which I'm hoping it does).  I don't like the `1` suffix though, due to potential confusion with `fold[lr]1`.  However, I think that adding those should be the job of a different PR than this one.

mb64 commented:
I rebased and updated the test output.  As for `fromList`, I'm choosing to leave it as is in this PR; I don't feel like I have the authority or expertise to decide its eventual fate.

gallais commented:
Fair enough. Thanks for the PR!
